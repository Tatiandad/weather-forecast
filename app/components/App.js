import React from 'react';
import Day from './Day/Day';
import CurrentWeather from './CurrentWeather/CurrentWeather'

const Api_Key = 'f2cec2774f3fcfa5572d11e08827b56f';

class App extends React.Component {
  state = {
    city: '',
    country: '',
    list: [],
    errorMessage: '',
    unit: 'C',
    currentWeather: {}
  }

  async getForecast(city, unit='F') {
    const tempUnit = unit === 'F' ? 'imperial' : 'metric';
    try {
      const api_call = await fetch(`https://api.openweathermap.org/data/2.5/forecast?q=${city}&appid=${Api_Key}&units=${tempUnit}`);
      const response = await api_call.json()
      if (response.cod === '200') {
        this.setState({ 
          list: response.list,
          country: response.city.country,
          city: response.city.name
        });
      } else {
        this.setState({ 
          errorMessage: response.message,
          city: '',
          country: '',
          list: [],
          unit: 'C',
          currentWeather: {}
        });
      }
    } catch(err) {
      console.log(err);
    }
  }

  async getTodaysWeather(city, unit='F') {
    const tempUnit = unit === 'F' ? 'imperial' : 'metric';
    try {
      const api_call = await fetch(`https://api.openweathermap.org/data/2.5/weather?q=${city}&appid=${Api_Key}&units=${tempUnit}`);
      const response = await api_call.json()
      if (response.cod === 200) {
        this.setState({
          currentWeather: {
            main: response.main,
            weather: response.weather,
            wind: response.wind
          }
        });
      } else {
        this.setState({ 
          errorMessage: response.message,
          city: '',
          country: '',
          list: [],
          unit: 'C',
          currentWeather: {}
        });
      }
    } catch(err) {
      console.log(err);
    }
  }

  processData = () => {
    const days = new Map();
    this.state.list.forEach((item) => {
      const day = this.getDate(item.dt_txt);
      if( !days[day] ) days[day] = []
      days[day].push(item)
    });
    return days;
  }

  getDate = (dateString) => {
    const regex = /([0-9]{4}-[0-9]{2}-[0-9]{2})/;
    return dateString.match(regex)[0];
  }

  handlechange = (e) => {
    this.setState({ 
      city: e.target.value,
      errorMessage: '',
      list: [],
      unit: 'C',
      currentWeather: {}
    });
  }

  handleClick = () => {
    this.getTodaysWeather(this.state.city);
    this.getForecast(this.state.city);
  }

  toggleUnit = () => {
    const unit = this.state.unit === 'F' ? 'C' : 'F';
    this.setState({ unit });
    this.getTodaysWeather(this.state.city, this.state.unit);
    this.getForecast(this.state.city, this.state.unit)
  }

  render () {
    const days = this.processData();
    const unit = this.state.unit === 'F' ? 'C' : 'F';
    return (
      <div className='container'>
        <div className='container-controls'>
          <input type='text' value={this.state.city} onChange={this.handlechange} />
          <button name='goButton' onClick={this.handleClick}>Go</button>
          <button name='tempButton' onClick={this.toggleUnit}>&deg;{this.state.unit}</button>
        </div>
        {this.state.currentWeather.weather &&
          <div>
            <CurrentWeather 
              city={this.state.city}
              country={this.state.country}
              currentWeather={this.state.currentWeather}
              unit={unit}
            />
          </div>
        }
        { this.state.list.length > 0 && 
          <div className='forecast-container'>
            {Object.keys(days).map((day, idx) => {
              return (
                <Day 
                  key={`${day}-${idx}`} 
                  data={days[day]} 
                  date={this.getDate(days[day][0].dt_txt)} 
                  unit={unit}
                />)
            })}
          </div>
        }
        { this.state.errorMessage && 
          <div>
            {this.state.errorMessage}
          </div>
        }
      </div>
    )
  }
}

export default App;