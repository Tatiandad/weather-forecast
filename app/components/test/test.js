import React, { Component } from 'react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import "babel-polyfill";
import fetch from 'isomorphic-fetch';
import { shallow, mount, render } from 'enzyme';
import App from '../App';
import forecast from '../../utils/dummy-forecast';
import weather from '../../utils/dummy-weather';

configure({ adapter: new Adapter() });

let shallowWrapper;
let mountWrapper;

beforeEach(() => {
  shallowWrapper = shallow(<App />);
  mountWrapper = mount(<App />);
});

describe('<App />', () => {
  it('renders <App /> component', () => {
    expect(shallowWrapper).toMatchSnapshot();
    const input = shallowWrapper.find('input');
    const buttons = shallowWrapper.find('button');
    expect(input.length).toBe(1);
    expect(buttons.length).toBe(2);
  });

  it('temp buttons toggle on click', () => {
    let tempButton = shallowWrapper.find('button[name="tempButton"]');
    expect(tempButton.text()).toEqual(`°${shallowWrapper.state().unit}`);

    const instance = shallowWrapper.instance();
    const getWeather = jest.spyOn(instance, 'getTodaysWeather');
    const getForecast = jest.spyOn(instance, 'getForecast');
    shallowWrapper.setState({ city: 'London' });

    instance.forceUpdate();

    tempButton.simulate('click');
    
    tempButton = shallowWrapper.find('button[name="tempButton"]');
    expect(tempButton.text()).toEqual(`°F`);
    expect(getWeather).toHaveBeenCalled();
    expect(getForecast).toHaveBeenCalled();
  });

  it('Go button triggers handleClick function', () => {
    const instance = shallowWrapper.instance();
    const handle = jest.spyOn(instance, 'handleClick');
    const ApiCall = jest.spyOn(instance, 'getForecast');

    instance.forceUpdate();    

    const goButton = shallowWrapper.find('button[name="goButton"]');
    goButton.simulate('click');
    expect(handle).toHaveBeenCalled();
    expect(ApiCall).toHaveBeenCalled();
  });

  it('Input change updates state', () => {
    const input = shallowWrapper.find('input');
    input.simulate('change', {target: { value: 'london' }});
    expect(shallowWrapper.state().city).toEqual('london');
  });

 it('Mocking forecast call success', () => {
    const getForecast = shallowWrapper.instance().getForecast;
    const fetchSpy = jest.spyOn(global, 'fetch')
      .mockImplementation(() => Promise.resolve({
        json: () => ( forecast ),
      }));
    
    return getForecast.call(shallowWrapper, 'London')
      .then(() => {
          expect(fetchSpy).toHaveBeenCalledWith('http://api.openweathermap.org/data/2.5/forecast?q=London&appid=f2cec2774f3fcfa5572d11e08827b56f&units=imperial');
          expect(shallowWrapper.state().list.length).toEqual(forecast.list.length);
      });
  });

  it('Mocking forecast call failure', () => {
    const getForecast = shallowWrapper.instance().getForecast;
    const fetchSpy = jest.spyOn(global, 'fetch')
      .mockImplementation(() => Promise.resolve({
        json: () => ({ cod:'404',message:'city not found' }),
      }));
    
    return getForecast.call(shallowWrapper, 'London')
      .then(() => {
          expect(fetchSpy).toHaveBeenCalledWith('http://api.openweathermap.org/data/2.5/forecast?q=London&appid=f2cec2774f3fcfa5572d11e08827b56f&units=imperial');
          expect(shallowWrapper.state().list.length).toEqual(0);
          expect(shallowWrapper.state().errorMessage).toEqual('city not found');
      });
  });
  
  it('Mocking weather call success', () => {
    const getTodaysWeather = shallowWrapper.instance().getTodaysWeather;
    const fetchSpy = jest.spyOn(global, 'fetch')
      .mockImplementation(() => Promise.resolve({
        json: () => ( weather ),
      }));
    
    return getTodaysWeather.call(shallowWrapper, 'London', 'C')
      .then(() => {
          expect(fetchSpy).toHaveBeenCalledWith('http://api.openweathermap.org/data/2.5/weather?q=London&appid=f2cec2774f3fcfa5572d11e08827b56f&units=metric');
          expect(shallowWrapper.state().currentWeather.main.temp).toEqual(weather.main.temp);
      });
  });
});