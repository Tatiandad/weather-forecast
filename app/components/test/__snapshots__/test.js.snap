// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders <App /> component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="container-controls"
        >
          <input
            onChange={[Function]}
            type="text"
            value=""
          />
          <button
            name="goButton"
            onClick={[Function]}
          >
            Go
          </button>
          <button
            name="tempButton"
            onClick={[Function]}
          >
            °
            C
          </button>
        </div>,
        undefined,
        false,
        "",
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              onChange={[Function]}
              type="text"
              value=""
            />,
            <button
              name="goButton"
              onClick={[Function]}
            >
              Go
            </button>,
            <button
              name="tempButton"
              onClick={[Function]}
            >
              °
              C
            </button>,
          ],
          "className": "container-controls",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Go",
              "name": "goButton",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Go",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "°",
                "C",
              ],
              "name": "tempButton",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              "°",
              "C",
            ],
            "type": "button",
          },
        ],
        "type": "div",
      },
      undefined,
      false,
      "",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="container-controls"
          >
            <input
              onChange={[Function]}
              type="text"
              value=""
            />
            <button
              name="goButton"
              onClick={[Function]}
            >
              Go
            </button>
            <button
              name="tempButton"
              onClick={[Function]}
            >
              °
              C
            </button>
          </div>,
          undefined,
          false,
          "",
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                onChange={[Function]}
                type="text"
                value=""
              />,
              <button
                name="goButton"
                onClick={[Function]}
              >
                Go
              </button>,
              <button
                name="tempButton"
                onClick={[Function]}
              >
                °
                C
              </button>,
            ],
            "className": "container-controls",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onChange": [Function],
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Go",
                "name": "goButton",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Go",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "°",
                  "C",
                ],
                "name": "tempButton",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                "°",
                "C",
              ],
              "type": "button",
            },
          ],
          "type": "div",
        },
        undefined,
        false,
        "",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
